src/librustdoc/passes/automatic_links.rs

```diff
@@ -0,0 +1,126 @@
+use super::{span_of_attrs, Pass};
+use crate::clean::*;
+use crate::core::DocContext;
+use crate::fold::DocFolder;
+use crate::html::markdown::opts;
+use core::ops::Range;
+use pulldown_cmark::{Event, LinkType, Parser, Tag};
+use regex::Regex;
+use rustc_errors::Applicability;
+use rustc_feature::UnstableFeatures;
+use rustc_session::lint;
+
+pub const CHECK_AUTOMATIC_LINKS: Pass = Pass {
+    name: "check-automatic-links",
+    run: check_automatic_links,
+    description: "detects URLS/email addresses that could be written using angle brackets",
+};
+
+const URL_REGEX: &str =
+    r"https?://([-a-zA-Z0-9@:%._\+~#=]{2,256}\.)+[a-zA-Z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)";
```

Comment:
This is a little hard to read.
```suggestion
const URL_REGEX: &str = concat!(
        r"https?://",  // url scheme
        r"([-a-zA-Z0-9@:%._\+~#=]{2,256}\.)+",  // one or more subdomains
        r"[a-zA-Z]{2,4}",  // root domain
        r"\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"  // optional query or url fragments
    );
```

Reply:
Also, I'm still not sure what `\b` does. Maybe instead make this more precise and change the query section to `(\?[-a-zA-Z0-9@:%_\+.~#?&//=]*)?(#[-a-zA-Z0-9@:%_\+.~#?&//=]*)?`
```

Reply:
Last comment: I think the `https://` regex will prevent you from picking up on email links. If you want to keep that in the docs, please add a test for `me@example.com` (but I'd be fine with you just saying email isn't supported, regexes for emails are _awful_).

Reply:
I don't really care about emails so let's not. :)